k2/csrc/array.h

```diff
@@ -49,6 +55,44 @@ class Array1 {
                                        byte_offset_);
   }
 
+  // Return a reference to this viewed as `Any` type (for when we
+  // want a generic array without type informatiopn)
+  // Reverse this with Specialize().
+  Array1<Any> &Generic() {
+    return *reinterpret_cast<Array1<Any>*>(this);
+  }
+  const Array1<Any> &Generic() const {
+    return *reinterpret_cast<const Array1<Any>*>(this);
+  }
+
+
+  // Only if this is a Array1<Any>: convert to a specific type, which must be
+  // identical to the actual type stored as dtype_.
+  // You call this as, e.g.
+  //  Array1<Any> a(...);
+  //  if (a.GetDtype() == kInt32Dtype) {
+  //     Array1<int32_t> &b = a.template Specialize<int32_t>();
+  //     ...
+  //  }
+  // Note: the '.template' above is only necessary if you are inside
+  // a template definition already.  It is an odd syntax.
+  template <typename U>
+  Array1<U> &Specialize() {
+    static_assert(std::is_same<T, Any>::value,
+                  "generic arrays not supported here");
+    K2_CHECK_EQ(dtype_, DtypeOf<U>::dtype);
+    return *reinterpret_cast<Array1<U>*>(this);
+  }
+  template <typename U>
+  Array1<U> &Specialize() const {
```

Comment:
Will this compile if you instantiate it?

The return value is a const lvalue, but the return type is an lvalue reference.